File: src/main/java/com/milaboratory/mixcr/trees/DebugInfo.java
package com.milaboratory.mixcr.trees;

import com.google.common.collect.ImmutableMap;
import com.milaboratory.core.Range;
import com.milaboratory.core.mutations.Mutations;
import com.milaboratory.core.sequence.NucleotideSequence;

import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public class DebugInfo {
    public static Map<String, Function<DebugInfo, Object>> COLUMNS_FOR_XSV = ImmutableMap.<String, Function<DebugInfo, Object>>builder()
            .put("VGeneName", it -> it.rootInfo.getVJBase().VGeneName)
            .put("JGeneName", it -> it.rootInfo.getVJBase().JGeneName)
            .put("CDR3Length", it -> it.rootInfo.getVJBase().CDR3length)
            .put("VRangeWithoutCDR3", it -> it.VRangesWithoutCDR3.stream().map(DebugInfo::encodeRange).collect(Collectors.joining(",")))
            .put("VRangeInCDR3", it -> encodeRange(it.rootInfo.getVRangeInCDR3()))
            .put("JRangeInCDR3", it -> encodeRange(it.rootInfo.getJRangeInCDR3()))
            .put("JRangeWithoutCDR3", it -> it.JRangesWithoutCDR3.stream().map(DebugInfo::encodeRange).collect(Collectors.joining(",")))
            .put("treeId", it -> it.treeId.getId())
            .put("cloneId", it -> it.cloneId)
            .put("id", it -> it.id)
            .put("parentId", it -> it.parentId)
            .put("NDN", it -> it.NDN)
            .put("VMutationsFromRoot", it -> it.mutationsFromRoot.VMutations.encode())
            .put("JMutationsFromRoot", it -> it.mutationsFromRoot.JMutations.encode())
            .put("NDNMutationsFromRoot", it -> it.mutationsFromRoot.NDNMutations.encode())
            .put("VMutationsFromParent", it -> it.mutationsFromParent != null ? it.mutationsFromParent.VMutations.encode() : null)
            .put("JMutationsFromParent", it -> it.mutationsFromParent != null ? it.mutationsFromParent.JMutations.encode() : null)
            .put("NDNMutationsFromParent", it -> it.mutationsFromParent != null ? it.mutationsFromParent.NDNMutations.encode() : null)
            .put("decisionMetric", it -> it.decisionMetric)
            .put("publicClone", it -> it.publicClone)
            .build();

    private static String encodeRange(Range range) {
        return range.getLower() + "-" + range.getUpper();
    }

    public static Range decodeRange(String raw) {
        String[] split = raw.split("-");
        return new Range(Integer.parseInt(split[0]), Integer.parseInt(split[1]));
    }

    private final TreeWithMetaBuilder.TreeId treeId;
    private final RootInfo rootInfo;
    private final List<Range> VRangesWithoutCDR3;
    private final List<Range> JRangesWithoutCDR3;
    private final Integer cloneId;
    private final int id;
    private final Integer parentId;
    private final NucleotideSequence NDN;
    private final MutationsSet mutationsFromRoot;
    private final MutationsSet mutationsFromParent;
    private final Double decisionMetric;
    private final boolean publicClone;


    public DebugInfo(
            TreeWithMetaBuilder.TreeId treeId,
            RootInfo rootInfo,
            List<Range> VRangesWithoutCDR3,
            List<Range> JRangesWithoutCDR3,
            Integer cloneId,
            int id,
            Integer parentId,
            NucleotideSequence NDN,
            MutationsSet mutationsFromRoot,
            MutationsSet mutationsFromParent,
            Double decisionMetric,
            boolean publicClone
    ) {
        this.treeId = treeId;
        this.rootInfo = rootInfo;
        this.VRangesWithoutCDR3 = VRangesWithoutCDR3;
        this.JRangesWithoutCDR3 = JRangesWithoutCDR3;
        this.cloneId = cloneId;
        this.id = id;
        this.parentId = parentId;
        this.NDN = NDN;
        this.mutationsFromRoot = mutationsFromRoot;
        this.mutationsFromParent = mutationsFromParent;
        this.decisionMetric = decisionMetric;
        this.publicClone = publicClone;
    }

    public static class MutationsSet {
        private final Mutations<NucleotideSequence> VMutations;
        private final Mutations<NucleotideSequence> NDNMutations;
        private final Mutations<NucleotideSequence> JMutations;

        public MutationsSet(
                Mutations<NucleotideSequence> VMutations,
                Mutations<NucleotideSequence> NDNMutations,
                Mutations<NucleotideSequence> JMutations
        ) {
            this.VMutations = VMutations;
            this.NDNMutations = NDNMutations;
            this.JMutations = JMutations;
        }
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/trees/DebugInfo.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.trees

import com.google.common.collect.ImmutableMap
import com.milaboratory.core.Range
import com.milaboratory.core.sequence.NucleotideSequence

class DebugInfo(
    private val treeId: TreeId,
    private val rootInfo: RootInfo,
    private val VRangesWithoutCDR3: Collection<Range>,
    private val JRangesWithoutCDR3: Collection<Range>,
    private val cloneIds: CloneWrapper.ID?,
    private val id: Int,
    private val parentId: Int?,
    private val NDN: NucleotideSequence,
    private val mutationsFromRoot: MutationsSet,
    private val mutationsFromParent: MutationsSet?,
    private val decisionMetric: Double?
) {
    companion object {
        var COLUMNS_FOR_XSV: Map<String, (DebugInfo) -> Any?> =
            ImmutableMap.builder<String, (DebugInfo) -> Any?>()
                .put("VGeneName") { it.rootInfo.VJBase.geneIds.V.name }
                .put("JGeneName") { it.rootInfo.VJBase.geneIds.J.name }
                .put("CDR3Length") { it.rootInfo.VJBase.CDR3length }
                .put("VRangeWithoutCDR3") {
                    it.VRangesWithoutCDR3.joinToString(",") { range -> encodeRange(range) }
                }
                .put("VRangeInCDR3") { encodeRange(it.rootInfo.rangeInCDR3.V) }
                .put("JRangeInCDR3") { encodeRange(it.rootInfo.rangeInCDR3.J) }
                .put("JRangeWithoutCDR3") {
                    it.JRangesWithoutCDR3.joinToString(",") { range -> encodeRange(range) }
                }
                .put("treeIdFull") { it.treeId.encode() }
                .put("treeId") { it.treeId.id }
                .put("clonesIds") { info -> info.cloneIds?.ids?.joinToString(",") { it.encode() } }
                .put("clonesCount") { it.cloneIds?.ids?.size ?: 0 }
                .put("id") { it.id }
                .put("parentId") { it.parentId }
                .put("NDN") { it.NDN }
                .put("VMutationsFromRoot") { it.mutationsFromRoot.mutations.V.combinedMutations().encode() }
                .put("JMutationsFromRoot") { it.mutationsFromRoot.mutations.J.combinedMutations().encode() }
                .put("NDNMutationsFromRoot") { it.mutationsFromRoot.NDNMutations.mutations.encode() }
                .put("VMutationsFromParent") { it.mutationsFromParent?.mutations?.V?.combinedMutations()?.encode() }
                .put("JMutationsFromParent") { it.mutationsFromParent?.mutations?.J?.combinedMutations()?.encode() }
                .put("NDNMutationsFromParent") { it.mutationsFromParent?.NDNMutations?.mutations?.encode() }
                .put("decisionMetric") { it.decisionMetric }
                .build()

        private fun encodeRange(range: Range): String = range.lower.toString() + "-" + range.upper

        fun decodeRange(raw: String): Range {
            val split = raw.split(Regex("-")).dropLastWhile { it.isEmpty() }.toTypedArray()
            return Range(split[0].toInt(), split[1].toInt())
        }
    }
}
